Replace gometalinter with golangci-lint

thaJeztah opened the pull request:
~rebased on top of https://github.com/moby/moby/pull/39902~

Replacing gometalinter, because it has been deprecated; https://github.com/alecthomas/gometalinter. Similar PR for the CLI; https://github.com/docker/cli/pull/1797

Quite a few issues were picked by the linters provided by golangci-lint, so unfortunately it's a big PR (some bugixes were already separated out)

thaJeztah committed:
Replace gometalinter with golangci-lint

Replacing gometalinter, because it has been deprecated,
and the repository was archived.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon/seccomp_linux.go: fix error-capitalization (golint)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: fix capitalization of variables and errors (golint)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon/cluster: fix unused context (staticcheck)

```
daemon/cluster/nodes.go:69:36: SA4009: argument ctx is overwritten before first use (staticcheck)
13:06:14 	return c.lockedManagerAction(func(ctx context.Context, state nodeState) error {
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
DriverBenchDiffApplyN: ignore empty branch (staticcheck)

suppressing the "SA9003: empty branch (staticcheck)" instead of commenting-out
or removing these lines because removing/commenting these lines causes a ripple
effect of changes, and there's still a to-do below.

```
13:06:14 daemon/graphdriver/graphtest/graphbench_unix.go:175:3: SA9003: empty branch (staticcheck)
13:06:14 		if applyDiffSize != diffSize {
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: fix incorrect permissions (staticcheck)

```
13:06:14 integration-cli/docker_api_containers_test.go:1983:72: SA9002: file mode '666' evaluates to 01232; did you mean '0666'? (staticcheck)
13:06:14 	err = ioutil.WriteFile(filepath.Join(tmpDir, "bar"), []byte("hello"), 666)
13:06:14 integration-cli/fixtures_linux_daemon_test.go:64:46: SA9002: file mode '600' evaluates to 01130; did you mean '0600'? (staticcheck)
13:06:14 	err = ioutil.WriteFile(dockerFile, content, 600)
13:06:14 	                                            ^
13:06:14 integration-cli/fixtures_linux_daemon_test.go:119:54: SA9002: file mode '600' evaluates to 01130; did you mean '0600'? (staticcheck)
13:06:14 	err = ioutil.WriteFile(dockerfile, []byte(content), 600)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
TestParseRelease: fix missing assert

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/devicemapper: disable unused and varcheck linters

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/pools/pools.go: suppress SA6002: argument should be pointer-like  for now

```
13:06:14 pkg/pools/pools.go:75:13: SA6002: argument should be pointer-like to avoid allocations (staticcheck)
13:06:14 	bp.pool.Put(b)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
TestMoveToSubdir: use sort.Strings() (gosimple)

```
13:06:14 pkg/directory/directory_test.go:182:2: S1032: should use sort.Strings(...) instead of sort.Sort(sort.StringSlice(...)) (gosimple)
13:06:14 	sort.Sort(sort.StringSlice(results))
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon.getEndpointInNetwork() is only used on Windows

```
13:06:14 daemon/network.go:964:6: U1000: func `getEndpointInNetwork` is unused (unused)
13:06:14 func getEndpointInNetwork(name string, n libnetwork.Network) (libnetwork.Endpoint, error) {
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
volume Create: fix incorrect file permissions (staticcheck)

```
14:01:54 volume/local/local.go:175:80: SA9002: file mode '600' evaluates to 01130; did you mean '0600'? (staticcheck)
14:01:54 		if err = ioutil.WriteFile(filepath.Join(filepath.Dir(path), "opts.json"), b, 600); err != nil {
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
cluster/executor: remove unused containerConfig.endpoint()

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
goimports: fix imports

Format the source according to latest goimports.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: use raw strings for regexes (gosimple)

```
14:26:43 integration-cli/docker_cli_build_test.go:3430:15: S1007: should use raw string (`...`) with regexp.MustCompile to avoid having to escape twice (gosimple)
14:26:43 	outRegexp := regexp.MustCompile("^(sha256:|)[a-z0-9]{64}\\n$")
14:26:43 	             ^
14:26:43 integration-cli/docker_cli_by_digest_test.go:26:20: S1007: should use raw string (`...`) with regexp.MustCompile to avoid having to escape twice (gosimple)
14:26:43 	pushDigestRegex = regexp.MustCompile("[\\S]+: digest: ([\\S]+) size: [0-9]+")
14:26:43 	                  ^
14:26:43 integration-cli/docker_cli_by_digest_test.go:27:20: S1007: should use raw string (`...`) with regexp.MustCompile to avoid having to escape twice (gosimple)
14:26:43 	digestRegex     = regexp.MustCompile("Digest: ([\\S]+)")
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: remove unused functions and enable disabled tests

```
14:26:43 integration-cli/docker_cli_pull_local_test.go:64:6: U1000: func `testConcurrentPullWholeRepo` is unused (unused)
14:26:43 func testConcurrentPullWholeRepo(c *check.C) {
14:26:43      ^
14:26:43 integration-cli/docker_cli_pull_local_test.go:111:31: U1000: func `(*DockerRegistrySuite).testConcurrentPullWholeRepo` is unused (unused)
14:26:43 func (s *DockerRegistrySuite) testConcurrentPullWholeRepo(c *check.C) {
14:26:43                               ^
14:26:43 integration-cli/docker_cli_pull_local_test.go:115:38: U1000: func `(*DockerSchema1RegistrySuite).testConcurrentPullWholeRepo` is unused (unused)
14:26:43 func (s *DockerSchema1RegistrySuite) testConcurrentPullWholeRepo(c *check.C) {
14:26:43                                      ^
14:26:43 integration-cli/docker_cli_pull_local_test.go:120:6: U1000: func `testConcurrentFailingPull` is unused (unused)
14:26:43 func testConcurrentFailingPull(c *check.C) {
14:26:43      ^
14:26:43 integration-cli/docker_cli_pull_local_test.go:142:31: U1000: func `(*DockerRegistrySuite).testConcurrentFailingPull` is unused (unused)
14:26:43 func (s *DockerRegistrySuite) testConcurrentFailingPull(c *check.C) {
14:26:43                               ^
14:26:43 integration-cli/docker_cli_pull_local_test.go:146:38: U1000: func `(*DockerSchema1RegistrySuite).testConcurrentFailingPull` is unused (unused)
14:26:43 func (s *DockerSchema1RegistrySuite) testConcurrentFailingPull(c *check.C) {

14:26:43 integration-cli/docker_utils_test.go:170:6: U1000: func `inspectImage` is unused (unused)
14:26:43 func inspectImage(c *check.C, name, filter string) string {
14:26:43      ^
14:26:43 integration-cli/events_utils_test.go:196:6: U1000: func `parseEventsWithID` is unused (unused)
14:26:43 func parseEventsWithID(c *check.C, out, match, id string) {
14:26:43      ^
14:26:43 integration-cli/fixtures_linux_daemon_test.go:17:6: U1000: type `testingT` is unused (unused)
14:26:43 type testingT interface {
14:26:43      ^
14:26:43 integration-cli/fixtures_linux_daemon_test.go:19:2: U1000: func `testingT.Fatalf` is unused (unused)
14:26:43 	Fatalf(string, ...interface{})
14:26:43 	^
14:26:43 integration-cli/fixtures_linux_daemon_test.go:22:6: U1000: type `logT` is unused (unused)
14:26:43 type logT interface {
14:26:43      ^
14:26:43 integration-cli/fixtures_linux_daemon_test.go:23:2: U1000: func `logT.Logf` is unused (unused)
14:26:43 	Logf(string, ...interface{})
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Fix and re-enable TestDockerNetworkMacvlan/MultiSubnet

The linter marked an issue because testMacvlanMultiSubnet was not used.
Re-enabling the test showed that there was a typo in the assert, causing the
test to fail:

```
--- FAIL: TestDockerNetworkMacvlan/MultiSubnet (4.74s)
    macvlan_test.go:243: assertion failed: 2001:db8:abc4::254 (c3.NetworkSettings.Networks["dualstackbridge"].IPv6Gateway string) != 2001:db8.abc4::254 (string)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration/network: remove unuset setupTest()

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: fix minor linting issues

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
volume/local fix file permissions

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Remove unused functions, variables, fields

opts/env_test: suppress a linter warning

this one:

> opts/env_test.go:95:4: U1000: field `err` is unused (unused)
> 			err      error
>			^

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
API: explicitely ignore unhandled errors

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: remove unnescessary conversions (unconvert)

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin committed:
pkg/term/term_linux_test: check returned error before deferring tty.Close() (staticcheck)

Fix the following warnings from staticcheck linter:

```
pkg/term/term_linux_test.go:34:2: SA5001: should check returned error before deferring tty.Close() (staticcheck)
	defer tty.Close()
	^
pkg/term/term_linux_test.go:52:2: SA5001: should check returned error before deferring tty.Close() (staticcheck)
 	defer tty.Close()
	^
pkg/term/term_linux_test.go:67:2: SA5001: should check returned error before deferring tty.Close() (staticcheck)
	defer tty.Close()
	^
....
```

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin committed:
builder/copy-unix: fix filepath.Walk args

Mark unused WalkFunc arguments as such, to get rid of the
following linterwarning:

> builder/dockerfile/copy_unix.go:28:71: SA4009: argument err is overwritten before first use (staticcheck)
> 	return filepath.Walk(source, func(fullpath string, info os.FileInfo, err error) error {

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
client: suppress filters.ToParamWithVersion lint warning

Add annotations to suppress warnings like this one:

> client/container_list.go:38:22: SA1019: filters.ToParamWithVersion is deprecated: Use ToJSON  (staticcheck)
> 		filterJSON, err := filters.ToParamWithVersion(cli.version, options.Filters)
>		                   ^

Modify the deprecation notice to specify it is applicable to new code
only.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
client: remove put()

Apparently it is not used anywhere

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah committed:
integration-cli: remove unused types and fields

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin committed:
daemon/monitor: rm redundant if

The last check for err != nil is not needed as err is always non-nil
there. Remove the check.

Also, no need to explicitly define `var err error` here.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
daemon/checkpoint: rm extra checks

In this code, err is already checked to be nil (or non-nil), so no need
to repeat extra checks.

Fixes the following govet warnings:

> daemon/checkpoint.go:38:12: nilness: tautological condition: nil == nil (govet)
> 		case err == nil:
> 		         ^
> daemon/checkpoint.go:45:12: nilness: tautological condition: nil == nil (govet)
> 		case err == nil && stat.IsDir():
> 		         ^
> daemon/checkpoint.go:47:12: nilness: tautological condition: nil == nil (govet)
> 		case err == nil:
> 		         ^

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
daemon/graphdriver/copy: rm dead code

Here, err is never non-nil as it was checked earlier.

Fixes the following linter warning:

> daemon/graphdriver/copy/copy.go:136:10: nilness: impossible condition: nil != nil (govet)
>		if err != nil {
>		       ^

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
integration-cli: TestBuildAddBadLinks, TestBuildAddBadLinksVolume: simplify check

This fixes the following lint warning:

> integration-cli/docker_cli_build_test.go:1001:59: nilness: tautological condition: non-nil != nil (govet)
>	if _, err := os.Stat(nonExistingFile); err == nil || err != nil && !os.IsNotExist(err) {
>	                                                         ^

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin committed:
TestDisplayJSONMessagesStreamInvalidJSON: fix error check

As pointed out by govet,

> pkg/jsonmessage/jsonmessage_test.go:231:94: nilness: nil dereference in dynamic method call (govet)
> 	if err := DisplayJSONMessagesStream(reader, data, inFd, false, nil); err == nil && err.Error()[:17] != "invalid character" {
>	                                                                                            ^

The nil deref never happened as err was always non-nil, and so the check
for error message text was not performed.

Fix this, and while at it, refactor the code a bit.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
plugin/setupNewPlugin: rm dead code

> plugin/manager_linux.go:285:9: nilness: impossible condition: nil != nil (govet)
>	if err != nil {
>	       ^

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
registry/newV1EndpointFromStr: rm dead code

> registry/endpoint_v1.go:127:9: nilness: impossible condition: nil != nil (govet)
> 	if err != nil {
>	       ^

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented on daemon/graphdriver/graphtest/graphbench_unix.go:
```diff
@@ -172,10 +172,10 @@ func DriverBenchDiffApplyN(b *testing.B, fileCount int, drivername string, drive
-			//b.Fatalf("Apply diff size different, got %d, expected %s", applyDiffSize, diffSize)
-		}
+		// TODO: enforce this
+		//if applyDiffSize != diffSize {
```
This removes one linter warning, but introduces another:
```
daemon/graphdriver/graphtest/graphbench_unix.go:167:3: applyDiffSize declared but not used (typecheck)
		applyDiffSize, err := driver.ApplyDiff(diff, "", arch)
		^
```

I guess the best way would be to use `// lint:ignore SA9003 because TODO` instead of commenting out the check.

kolyshkin commented on pkg/loopback/loop_wrapper.go:
```diff
@@ -16,6 +16,9 @@ package loopback // import "github.com/docker/docker/pkg/loopback"
 
+// nolint: deadcode
+// nolint: unused
+// nolint: varcheck
```
Apparently this is still not good enough :sob: as structcheck also complains about unused fields.

Also, to make things more compact, maybe use `// nolint: deadcode,unused,varcheck,structcheck`

thaJeztah replied:
Ah, good one; was actually wondering if they could be combined in a single line 👍

kolyshkin committed:
Fix some inefassign warnings

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
jsonfilelog_test: check err from open before close

Fix warnings like this one:

> daemon/logger/jsonfilelog/jsonfilelog_test.go:191:3: SA5001: should check returned error before deferring file.Close() (staticcheck)
> 		defer file.Close()
>		^

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
Use io.SeekStart instead of os.SEEK_SET

Fixes linter warnings like this one:

> distribution/pull_v2.go:229:39: SA1019: os.SEEK_SET is deprecated: Use io.SeekStart, io.SeekCurrent, and io.SeekEnd.  (staticcheck)

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah committed:
golangci-lint: enable gosec linter

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Builder: remove unused field

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/signal: remove unnescessary conversion (unconvert)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
TestUlimitOpt: fix composite literal uses unkeyed fields (govet)

```
18:15:45 opts/ulimit_test.go:11:13: composites: `*github.com/docker/docker/vendor/github.com/docker/go-units.Ulimit` composite literal uses unkeyed fields (govet)
18:15:45 		"nofile": {"nofile", 1024, 512},
18:15:45 		          ^
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/ioutils use sha256 instead of sha1 in tests

Not really important as it's just test-code, but prevents
gosec linting from producing warnings

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin committed:
TestTransfer*: don't call t.Fatal from a goroutine

staticcheck go linter warns:

> distribution/xfer/transfer_test.go:37:2: SA2002: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (staticcheck)

What it doesn't say is why. The reason is, t.Fatalf() calls t.FailNow(),
which is expected to stop test execution right now. It does so by
calling runtime.Goexit(), which, unless called from a main goroutine,
does not stop test execution.

Anyway, long story short, if we don't care much about stopping the test
case immediately, we can just replace t.Fatalf() with t.Errorf() which
still marks the test case as failed, but won't stop it immediately.

This patch was tested to check that the test fails if any of the
goroutines call t.Errorf():

1. Failure in DoFunc ("transfer function not started ...") was tested by
decreading the NewTransferManager() argument:

-        tm := NewTransferManager(5)
+        tm := NewTransferManager(2)

2. Failure "got unexpected progress value" was tested by injecting a random:

-                       if present && p.Current <= val {
+                       if present && p.Current <= val || rand.Intn(100) > 80 {

3. Failure in DoFunc ("too many jobs running") was tested by increasing
the NewTransferManager() argument:

-       tm := NewTransferManager(concurrencyLimit)
+       tm := NewTransferManager(concurrencyLimit + 1)

While at it:
 * fix/amend some error messages
 * use _ for unused arguments of DoFunc

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
daemon/logger/BenchmarkCopy: don't call t.Fatal from a goroutine

staticcheck go linter says:

> daemon/logger/copier_test.go:451:2: SA2002: the goroutine calls T.Fatal, which must be called in the same goroutine as the test (staticcheck)

What it doesn't say is why. The reason is, t.Fatal() calls t.FailNow(),
which is expected to stop test execution right now. It does so by
calling runtime.Goexit(), which, unless called from a main goroutine,
does not stop test execution.

Anyway, long story short, if we don't care much about stopping the test
case immediately, we can just replace t.Fatalf() with t.Errorf() which
still marks the test case as failed, but won't stop it immediately.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah committed:
golangci-lint: set correct build-tags to lint journald if supported

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/term/windows: add missing build-tag

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin committed:
distribution/xfer/download_test: rm unused field

> distribution/xfer/download_test.go:29:2: `os` is unused (structcheck)

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah committed:
unconvert: remove unnescessary conversions

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin committed:
builder-next/patchImageConfig: nitpicks

1. Do not introduce another dt variable.

2. Wrap the error similar to all the others.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

thaJeztah commented:
These powerpc failures look related; https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/15261/console

```
02:37:58 === Failed
02:37:58 === FAIL: builder/fscache TestFSCache (0.03s)
02:37:58 panic: interface conversion: interface {} is *[]uint8, not []uint8 [recovered]
02:37:58 	panic: interface conversion: interface {} is *[]uint8, not []uint8

02:37:58 === FAIL: daemon/graphdriver/copy TestCopyWithoutRange (0.02s)
02:37:58 panic: interface conversion: interface {} is *[]uint8, not []uint8 [recovered]
02:37:58 	panic: interface conversion: interface {} is *[]uint8, not []uint8

02:37:58 === FAIL: daemon/graphdriver/overlay TestOverlay50LayerRead (0.09s)
02:37:58 panic: interface conversion: interface {} is *[]uint8, not []uint8 [recovered]
02:37:58 	panic: interface conversion: interface {} is *[]uint8, not []uint8

02:37:58 === FAIL: daemon/logger/splunk TestInlineFormatWithNonDefaultOptions (0.07s)
02:37:58 panic: interface conversion: interface {} is *[]uint8, not []uint8 [recovered]
02:37:58 	panic: interface conversion: interface {} is *[]uint8, not []uint8

02:37:58 === FAIL: pkg/archive TestReplaceFileTarWrapper (1.12s)
02:37:58 panic: interface conversion: interface {} is *[]uint8, not []uint8

02:37:58 === FAIL: pkg/filenotify TestPollerEvent (0.80s)
02:37:58     poller_test.go:75: timeout waiting for event CHMOD
```

thaJeztah committed:
govet: suppress warning about Parent field that repeated json tag "parent"

```
image/image.go:65:2: structtag: struct field Parent repeats json tag "parent" also at image.go:39 (govet)
	Parent     ID        `json:"parent,omitempty"`
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/containerfs: rename output variable to prevent shadowing (govet)

```
pkg/containerfs/archiver.go:121:6: shadow: declaration of "err" shadows declaration at line 92 (govet)
	if err := dstDriver.MkdirAll(dstDriver.Dir(dst), 0700); err != nil {
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on image/image.go:
```diff
@@ -62,7 +62,6 @@ type V1Image struct {
 // Image stores the image configuration
 type Image struct {
 	V1Image
-	Parent     ID        `json:"parent,omitempty"`
```
That didn't work;

```
[2019-08-27T17:09:04.649Z] # github.com/docker/docker/image
[2019-08-27T17:09:04.649Z] image/store.go:217:18: cannot assign ID to img.V1Image.Parent (type string) in multiple assignment
```

thaJeztah committed:
logger/journald: U1000: field `mu` is unused (unused)

this looks to be a false positive, but this field is not
used if journald is not supported, which may be the cause

```
daemon/logger/journald/journald.go:21:2: U1000: field `mu` is unused (unused)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
gosec: add ignore comments for reported issues that can be ignored

```
builder/remotecontext/remote.go:48:        G107: Potential HTTP request made with variable url (gosec)
builder/remotecontext/git/gitutils.go:145: G107: Potential HTTP request made with variable url (gosec)
builder/remotecontext/git/gitutils.go:147: G107: Potential HTTP request made with variable url (gosec)
pkg/fileutils/fileutils_test.go:185:       G303: File creation in shared tmp directory without using ioutil.Tempfile (gosec)
pkg/tarsum/tarsum_test.go:7:               G501: Blacklisted import `crypto/md5`: weak cryptographic primitive (gosec)
pkg/tarsum/tarsum_test.go:9:               G505: Blacklisted import `crypto/sha1`: weak cryptographic primitive (gosec)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
structcheck: nolint for false positives

```
builder/fscache/fscache.go:505:2: `src` is unused (structcheck)
builder/fscache/fscache.go:507:2: `cached` is unused (structcheck)

daemon/graphdriver/vfs/quota_linux.go:9:2: `quotaCtl` is unused (structcheck)
daemon/graphdriver/vfs/quota_linux.go:10:2: `quotaOpt` is unused (structcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
TestWalkValues: add missing error-check (errcheck)

```
api/types/filters/parse_test.go:340:14: Error return value of `f.WalkValues` is not checked (errcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
api/server: explicitly ignore unhandled errors (errcheck)

```
api/server/router/build/build_routes.go:309:41: Error return value of `(*encoding/json.Decoder).Decode` is not checked (errcheck)
api/server/router/build/build_routes.go:431:11: Error return value of `io.Copy` is not checked (errcheck)
api/server/router/container/container_routes.go:582:13: Error return value of `conn.Write` is not checked (errcheck)
api/server/router/grpc/grpc_routes.go:38:12: Error return value of `conn.Write` is not checked (errcheck)
api/server/router/grpc/grpc_routes.go:39:12: Error return value of `resp.Write` is not checked (errcheck)
api/server/router/image/image_routes.go:94:15: Error return value of `output.Write` is not checked (errcheck)
api/server/router/image/image_routes.go:139:15: Error return value of `output.Write` is not checked (errcheck)
api/server/router/image/image_routes.go:164:15: Error return value of `output.Write` is not checked (errcheck)
api/server/router/image/image_routes.go:180:15: Error return value of `output.Write` is not checked (errcheck)
api/server/router/plugin/plugin_routes.go:126:15: Error return value of `output.Write` is not checked (errcheck)
api/server/router/plugin/plugin_routes.go:165:15: Error return value of `output.Write` is not checked (errcheck)
api/server/router/plugin/plugin_routes.go:273:15: Error return value of `output.Write` is not checked (errcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
builder-next: explicitely ignore unhandled errors (errcheck)

```
builder/builder-next/adapters/containerimage/pull.go:278:23: Error return value of `resolveProgressDone` is not checked (errcheck)
builder/builder-next/adapters/containerimage/pull.go:286:24: Error return value of `resolveProgressDone` is not checked (errcheck)
builder/builder-next/adapters/containerimage/pull.go:303:24: Error return value of `resolveProgressDone` is not checked (errcheck)
builder/builder-next/adapters/containerimage/pull.go:309:24: Error return value of `resolveProgressDone` is not checked (errcheck)
builder/builder-next/adapters/containerimage/pull.go:315:22: Error return value of `resolveProgressDone` is not checked (errcheck)
builder/builder-next/adapters/containerimage/pull.go:512:14: Error return value of `pw.Write` is not checked (errcheck)
builder/builder-next/adapters/containerimage/pull.go:675:13: Error return value of `pw.Write` is not checked (errcheck)
builder/builder-next/adapters/containerimage/pull.go:700:14: Error return value of `pw.Write` is not checked (errcheck)
builder/builder-next/adapters/containerimage/pull.go:786:10: Error return value of `pw.Write` is not checked (errcheck)
builder/builder-next/adapters/containerimage/pull.go:791:11: Error return value of `pw.Write` is not checked (errcheck)

builder/builder-next/exporter/export.go:133:13: Error return value of `layersDone` is not checked (errcheck)
builder/builder-next/exporter/export.go:163:12: Error return value of `configDone` is not checked (errcheck)
builder/builder-next/exporter/export.go:172:11: Error return value of `tagDone` is not checked (errcheck)
builder/builder-next/exporter/writer.go:207:10: Error return value of `pw.Write` is not checked (errcheck)
builder/builder-next/exporter/writer.go:212:11: Error return value of `pw.Write` is not checked (errcheck)
builder/builder-next/worker/worker.go:397:6: Error return value of `done` is not checked (errcheck)
builder/builder-next/worker/worker.go:446:10: Error return value of `pw.Write` is not checked (errcheck)
builder/builder-next/worker/worker.go:451:11: Error return value of `pw.Write` is not checked (errcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
golangci-lint: enable misspell, tweak configuration

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
misspell: fix some spelling issues

```
internal/test/environment/environment.go:37:23: `useing` is a misspelling of `using`(misspell)
integration/container/wait_test.go:49:9: `waitres` is a misspelling of `waiters`(misspell)
integration/container/wait_test.go:95:9: `waitres` is a misspelling of `waiters`(misspell)
integration-cli/docker_api_containers_test.go:1042:7: `waitres` is a misspelling of `waiters`(misspell)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
cluster/controllers/plugin: remove unused Controller.taskID (unused)

```
daemon/cluster/controllers/plugin/controller.go:37:2: U1000: field `taskID` is unused (unused)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
oci: fix SA4009: argument e is overwritten before first use (staticcheck)

```
oci/devices_linux.go:64:72: SA4009: argument e is overwritten before first use (staticcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/devicemapper ignore SA4000 false positive (staticcheck)

```
pkg/devicemapper/devmapper_wrapper.go:209:206: SA4000: identical expressions on the left and right side of the '==' operator (staticcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: fix TestPushMultipleTags

This test was meant to _push_ images, not _pull_ them

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: SA5001: should check returned error before deferring reg.Close() (staticcheck)

```
integration-cli/docker_cli_registry_user_agent_test.go:78:2: SA5001: should check returned error before deferring reg.Close() (staticcheck)
integration-cli/docker_cli_v2_only_test.go:30:2:             SA5001: should check returned error before deferring reg.Close() (staticcheck)
integration-cli/docker_api_containers_test.go:392:3:         SA5001: should check returned error before deferring resp.Body.Close() (staticcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
cmd/dockerd: explicitly ignore unchecked errors (errcheck)

```
cmd/dockerd/config.go:50:18: Error return value of `flags.MarkHidden` is not checked (errcheck)
cmd/dockerd/config.go:55:22: Error return value of `flags.MarkDeprecated` is not checked (errcheck)
cmd/dockerd/config.go:78:18: Error return value of `flags.MarkHidden` is not checked (errcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: S1025: the argument is already a string (gosimple)

```
integration-cli/docker_cli_daemon_test.go:1753:32: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_daemon_test.go:1783:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_daemon_test.go:1893:92: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_external_volume_driver_test.go:444:34: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_external_volume_driver_test.go:600:36: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_external_volume_driver_test.go:602:36: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_external_volume_driver_test.go:610:34: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_external_volume_driver_test.go:613:34: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_external_volume_driver_test.go:614:36: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_external_volume_driver_test.go:617:36: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_plugins_test.go:431:39: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:174:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1046:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1071:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1074:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1079:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1087:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1102:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1108:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1128:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1323:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1329:32: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1388:34: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
integration-cli/docker_cli_swarm_test.go:1985:31: S1025: the argument is already a string, there's no need to use fmt.Sprintf (gosimple)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: TestAttachMultipleAndRestart: don't call t.Fatal from a goroutine

```
integration-cli/docker_cli_attach_test.go:44:: SA2002: the goroutine calls T.Fatal, which must be called in the same goroutine as the test (staticcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
client/hijack: suppress SA1019: httputil.ErrPersistEOF is deprecated (staticcheck)

Keeping this code for now to allow connecting to old daemons, but we might
want to remove this at some point

```
client/hijack.go:90:12: SA1019: httputil.ErrPersistEOF is deprecated: No longer used.  (staticcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
TestGraphdriverPluginV2: SA5001: should check returned error before deferring (staticcheck)

```
integration/plugin/graphdriver/external_test.go:427:2: SA5001: should check returned error before deferring responseReader.Close() (staticcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
golangci-lint: suppress Xattrs is deprecated: Use PAXRecords instead

```
pkg/archive/archive.go:407:3:       SA1019: hdr.Xattrs is deprecated: Use PAXRecords instead.  (staticcheck)
pkg/archive/archive.go:408:3:       SA1019: hdr.Xattrs is deprecated: Use PAXRecords instead.  (staticcheck)
pkg/archive/archive.go:661:26:      SA1019: hdr.Xattrs is deprecated: Use PAXRecords instead.  (staticcheck)
pkg/archive/archive_linux.go:47:7:  SA1019: hdr.Xattrs is deprecated: Use PAXRecords instead.  (staticcheck)
pkg/archive/archive_linux.go:48:12: SA1019: hdr.Xattrs is deprecated: Use PAXRecords instead.  (staticcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
golangci-lint: suppress remaining deprecation errors for now (staticcheck)

```
distribution/registry.go:84:3: SA1019: base.Dial is deprecated: Use DialContext instead, which allows the transport to cancel dials as soon as they are no longer needed. If both are set, DialContext takes priority.  (staticcheck)
registry/registry.go:188:3:    SA1019: base.Dial is deprecated: Use DialContext instead, which allows the transport to cancel dials as soon as they are no longer needed. If both are set, DialContext takes priority.  (staticcheck)

client/hijack.go:85:16:                               SA1019: httputil.NewClientConn is deprecated: Use the Client or Transport in package net/http instead.  (staticcheck)
integration-cli/docker_api_attach_test.go:245:12:     SA1019: httputil.NewClientConn is deprecated: Use the Client or Transport in package net/http instead.  (staticcheck)
integration/plugin/authz/authz_plugin_test.go:180:7:  SA1019: httputil.NewClientConn is deprecated: Use the Client or Transport in package net/http instead.  (staticcheck)
integration/plugin/authz/authz_plugin_test.go:479:12: SA1019: httputil.NewClientConn is deprecated: Use the Client or Transport in package net/http instead.  (staticcheck)

integration-cli/docker_api_attach_test.go:239:134:    SA1019: httputil.ClientConn is deprecated: Use Client or Transport in package net/http instead.  (staticcheck)

daemon/daemon.go:885:3:                               SA1019: grpc.WithDialer is deprecated: use WithContextDialer instead.  Will be supported throughout 1.x.  (staticcheck)

pkg/authorization/response.go:18:2:                   SA1019: http.CloseNotifier is deprecated: the CloseNotifier interface predates Go's context package. New code should use Request.Context instead.  (staticcheck)
pkg/authorization/response.go:160:30:                 SA1019: http.CloseNotifier is deprecated: the CloseNotifier interface predates Go's context package. New code should use Request.Context instead.  (staticcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
Revert "Revert "hack: temporarily disable deprecate-integration-cli""

This reverts commit 01332b54e02895c68b0ea3e4e1ff9d5dd8a0302a.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
reverting https://github.com/moby/moby/pull/39897 again for now

thaJeztah commented on pkg/filenotify/poller_test.go:
```diff
@@ -68,7 +68,7 @@ func TestPollerEvent(t *testing.T) {
 	}
 
-	if err := os.Chmod(f.Name(), 600); err != nil {
+	if err := os.Chmod(f.Name(), 0600); err != nil {
```
interesting; looks like this broke the test?

```
[2019-09-11T23:12:15.200Z] === [31mFailed[0m
[2019-09-11T23:12:15.200Z] === [31mFAIL[0m: pkg/filenotify TestPollerEvent (0.80s)
[2019-09-11T23:12:15.200Z]     poller_test.go:75: timeout waiting for event CHMOD

thaJeztah committed:
TestPollerEvent: fix filemode (staticcheck)

Staticcheck reported:

    SA9002: file mode '600' evaluates to 01130; did you mean '0600'? (staticcheck)

But fixing that caused the test to fail:

    === Failed
    === FAIL: pkg/filenotify TestPollerEvent (0.80s)
        poller_test.go:75: timeout waiting for event CHMOD

The problem turned out to be that the file was created with `0644`. However,
after umask, the file created actually had `0600` filemode. Running the `os.Chmod`
with `0600` therefore was a no-op, causing the test to fail (because no
CHMOD event would fire).

This patch changes the test to;

- create the file with mode `0600`
- assert that the file has the expected mode
- change the chmod to `0644`
- assert that it has the correct mode, before testing the event.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration: remove unused constants and fields (unused)

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
runconfig/config_test.go: remove unused test-utilities

The tests using these functions were removed in e89b6e8c2d2c36c43f22aeaf2a885646c2994051

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

kolyshkin commented on daemon/graphdriver/vfs/quota_linux.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+//nolint:structcheck
```
Maybe it makes sense to report this false positive upstream

kolyshkin commented on pkg/devicemapper/devmapper_wrapper.go:
```diff
@@ -206,6 +206,7 @@ func dmGetNextTargetFct(task *cdmTask, next unsafe.Pointer, start, length *uint6
 	}()
 
+	//lint:ignore SA4000 false positive on (identical expressions on the left and right side of the '==' operator) (staticcheck)
 	nextp := C.dm_get_next_target((*C.struct_dm_task)(task), next, &Cstart, &Clength, &CtargetType, &Cparams)
```
Spent 5 minutes looking at this and I could not imagine how the checker could come up seeing `==` in here :)

thaJeztah replied:
Could it be that detection devicemapper package is problematic because of CGO? 

We can create a tracking issue of these false positives (I assume they want a minimal reproducer, which might take some time to construct?)

kolyshkin commented on daemon/graphdriver/quota/errors.go:
```diff
@@ -3,6 +3,7 @@ package quota // import "github.com/docker/docker/daemon/graphdriver/quota"
 
 var (
+	//nolint
 	_ errdefs.ErrNotImplemented = (*errQuotaNotSupported)(nil)
```
This should be reported upstream

thaJeztah replied:
Good one; the `nolint` in the end didn't really help I think, so I think I should remove this commit, and create a tracking issue to investigate instead.

I have a feeling it might be due to a combination of build-tags and this anonymous import, but perhaps it's a bug in the linter indeed.

I'll drop this commit

kolyshkin commented on pkg/filenotify/poller_test.go:
```diff
@@ -68,6 +68,14 @@ func TestPollerEvent(t *testing.T) {
+	//
+	// === Failed
+	// === FAIL: pkg/filenotify TestPollerEvent (0.80s)
+	//     poller_test.go:75: timeout waiting for event CHMOD
```
I guess it fails because the file already had `0600` perm bits, so chmod has no effect and the event is not generated. Please use say `0640` (aka `-rw-r------`)

thaJeztah replied:
Right, so that's what I thought as well, but then looked at the line above, where the file is created with `0644`. https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/pkg/filenotify/poller_test.go#L64-L73

Perhaps it's because due to umask? haven't checked that one

thaJeztah replied:
Yup, looks to be umask, so the file is _created_ with `0600` (added an assert)

```
=== Failed
=== FAIL: pkg/filenotify TestPollerEvent (0.00s)
    poller_test.go:67: expected file /tmp/test-poller830233858 to have mode 0644, but got 0600
```

kolyshkin commented on hack/validate/golangci-lint.yml:
```diff
@@ -39,3 +39,9 @@ issues:
     - text: "G202: SQL string concatenation"
       linters:
         - gosec
+    - text: "SA1019: h.Xattrs is deprecated: Use PAXRecords instead"
```
this might be a legit warning... I started to take a look at it and got carried away :(

Anyway, let's address it as a followup.

thaJeztah replied:
Yes, I think all (or most) of the ones I'm ignoring in this file are legit, and things that should be addressed. I want to create tracking issues for those.

😂 I also started working on the `PAXRecords` one, but wasn't sure if my approach was right and/or if it would cause backward-incompatibility issues if we switched; so I removed those changes, and thought it was better left for a separate PR (and discussion)

kolyshkin reviewed (changes requested):
LGTM except for https://github.com/moby/moby/pull/39668/commits/ffb7d550ad25aef3d1bdebc73a24e0b48aabdf59 (please see my comment)

thaJeztah commented:
only failure is on RS1

```
540Z]     --- FAIL: Test/DockerSuite/TestStartReturnCorrectExitCode (9.91s)
[2019-09-13T15:19:50.540Z]         docker_cli_start_test.go:198: assertion failed: 125 (exitCode int) != 12 (int): out: time="2019-09-13T15:18:31Z" level=error msg="error waiting for container: EOF" 
[2019-09-13T15:19:50.540Z]             
[2019-09-13T15:19:50.540Z]         check_test.go:123: assertion failed: error is not nil: Error response from daemon: container 3f092576637bab6006d803e8c91cef8ac6ad9f4a6cff2ade3630bbe6d6019880: driver "windowsfilter" failed to remove root filesystem: hcsshim::GetComputeSystems: Access is denied.: failed to remove 3f092576637bab6006d803e8c91cef8ac6ad9f4a6cff2ade3630bbe6d6019880
[2019-09-13T15:19:50.540Z]

thaJeztah commented:
Argh, and now RS5 is failing on the same one;

```

[2019-09-13T22:49:34.539Z]     --- FAIL: Test/DockerSuite/TestStartReturnCorrectExitCode (4.05s)
[2019-09-13T22:49:34.539Z]         docker_cli_start_test.go:193: assertion failed: expected an error, got nil
```

kolyshkin reviewed (approved):
LGTM

kolyshkin commented:
Looks like this needs a rebase after #39715 got merged; otherwise ready to go!

thaJeztah commented:
rebased 👍

yongtang reviewed (approved):
LGTM

thaJeztah commented:
🥳🥳
